src/m3ninx/postings/roaring/bitmap_readonly_multi.go

```diff
@@ -0,0 +1,730 @@
+	multiContainerOpNegate
+)
+
+type containerIterator interface {
+	NextContainer() bool
+	ContainerKey() uint64
+	ContainerUnion(ctx containerOpContext, target *bitmapContainer)
+	ContainerIntersect(ctx containerOpContext, target *bitmapContainer)
+	ContainerNegate(ctx containerOpContext, target *bitmapContainer)
+}
+
+type containerOpContext struct {
+	// siblings is how many other containers at this container there is
+	// being operated on.
+	siblings int
+	// tempBitmap is useful for temporary scratch operations and allows
+	// for all sub-operations to share it rather than one per underlying
+	// container iterator.
+	tempBitmap *bitmapContainer
+}
+
+func newMultiBitmapIterator(
+	opts multiBitmapOptions,
+) *multiBitmapIterator {
+	var (
+		n     = len(opts.union) + len(opts.intersect) + len(opts.intersectNegate)
+		iters = make([]containerIteratorAndOp, 0, n)
+	)
+	iters = appendContainerItersWithOp(iters, opts.union, multiContainerOpUnion)
+	iters = appendContainerItersWithOp(iters, opts.intersect, multiContainerOpIntersect)
+	iters = appendContainerItersWithOp(iters, opts.intersectNegate, multiContainerOpNegate)
+	i := &multiBitmapIterator{
+		multiBitmapOptions: opts,
+		iters:              iters,
+		bitmap:             getBitmapContainer(),
+		tempBitmap:         getBitmapContainer(),
+	}
+	i.bitmapIter.Reset(0, i.bitmap)
+	return i
+}
+
+func appendContainerItersWithOp(
+	slice []containerIteratorAndOp,
+	iterables []multiBitmapIterable,
+	op multiContainerOp,
+) []containerIteratorAndOp {
+	for _, elem := range iterables {
+		var it containerIterator
+		switch {
+		case elem.multiBitmap != nil:
+			it = elem.multiBitmap.containerIterator()
+
+		case elem.bitmap != nil:
+			it = elem.bitmap.containerIterator()
+		}
+
+		if !it.NextContainer() {
+			continue
+		}
+
+		slice = append(slice, containerIteratorAndOp{
+			it: it,
+			op: op,
+		})
+	}
+	return slice
+}
+
+func (i *multiBitmapIterator) Next() bool {
+	for !i.bitmapIter.Next() {
+		// Reset to next containers.
+		var ok bool
+		i.iters, ok = i.multiContainerIter.resetAndReturnValid(i.iters)
+		if !ok {
+			// Entirely exhausted valid iterators.
+			return false
+		}
+
+		// Combine all current containers into single bitmap.
+		switch i.op { // Op is already validated at creation time.
+		case multiBitmapOpUnion:
+			// Start bitmap as unset.
+			i.bitmap.Reset(false)
+
+			// All are unions.
+			unions := i.filter(i.multiContainerIter.containerIters, multiContainerOpUnion)
+			ctx := containerOpContext{
+				siblings:   len(unions) - 1,
+				tempBitmap: i.tempBitmap,
+			}
+			for _, iter := range unions {
+				iter.it.ContainerUnion(ctx, i.bitmap)
+			}
+		case multiBitmapOpIntersect:
+			// Start bitmap as set, guaranteed to have one intersect call.
+			i.bitmap.Reset(true)
+
+			intersects := i.filter(i.multiContainerIter.containerIters, multiContainerOpIntersect)
+			negates := i.filter(i.multiContainerIter.containerIters, multiContainerOpNegate)
```

Comment:
Doesn't this overwrite the `i.filtered` in `intersects`?

Reply:
ah yes, TY!